### ISSUE TITLE
mxnet as backend

### ISSUE THREAD

**Issue by ares128** [fchollet/keras#1313]
as tensorflow already being added as a backend, can we add mxnet as a new backend?
it seems mxnet is faster than theano and has a smaller memory usage!

**Comment by fchollet** [comment]
MXNet is cool, but it doesn't look like it would be good match to Keras. What's the rationale here? "It's fast" doesn't cut it.

**Comment by gwding** [comment]
@fchollet could you explain a bit more on why MXNet is less a good match than TensorFlow to Keras? would you accept a MXNet backend PR if someone submit it?

**Comment by zachmayer** [comment]
I think mxnet implements both a front-end and a backend.  The backend could probably be hooked up to the keras frontend.

**Comment by fivejjs** [comment]
Sounds nice.

**Comment by bobchennan** [comment]
I think MXNet is a good choice because it supports multiple gpus and multiple nodes.

**Comment by EderSantana** [comment]
the problem is the API. if you can figure a way to write to the mxnet backend in a way that has autodiff, lazy graph compilation, etc (just like tensorflow and theano) it would be possible. In other words its not just a matter of writing all the layers in Keras with MXNet it has to fit Keras backend API. It was doable because Tensorflow and Theano have similar guiding principles

**Comment by fchollet** [comment]
The big question, of course, is who will write this new backend. If you can
write it or create a team that will, then go for it.

On Nov 29, 2016 10:34 AM, "Eder Santana" <notifications@github.com> wrote:

the problem is the API. if you can figure a way to write to the mxnet
backend in a way that has autodiff, lazy graph compilation, etc (just like
tensorflow and theano) it would be possible. In other words its not just a
matter of writing all the layers in Keras with MXNet it has to fit Keras
backend API. It was doable because Tensorflow and Theano have similar
guiding principles

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/fchollet/keras/issues/1313#issuecomment-263657240>, or mute
the thread

**Comment by EderSantana** [comment]
yeah, that is what I mean @fchollet . It will be hard work. But if they are really interested in getting people to do that. I'm sure the community will be happy with them.

**Comment by pGit1** [comment]
"I think MXNet is a good choice because it supports multiple gpus and multiple nodes."

This is the most important think Keras is missing IMHO. Keras should have a simple interface that lets you specify GPU IDs upfront and let it handle the rest of the backend stuff. This is a huge advantage that MXNET has and probably the main reason to make the switch from Keras to MXNET. No use having a powerful multi-gpu box when you can only use one GPU.

**Comment by EderSantana** [comment]
@pGit1 there is #3582 for that as well.

**Comment by pGit1** [comment]
@EderSantana I will checkout the link. I am not advocating for another back end by the way. I just want to utilize the full power of GPU box without worrying about GPU workload assignments and memory usage.

**Comment by fchollet** [comment]
Multi-GPU support will improve drastically in the near future. However,
this support will largely target TF only.

**Comment by pGit1** [comment]
`will largely target TF only.`

Well I better rewrite all my code with TF backend (kind of sux I hear people complain about how RNNs are much better on Theano). Thankfully I think the only think the only thing that will change for me is image input dimensions for Conv layers.

**Comment by fchollet** [comment]
If you are using Keras then you don't need to "rewrite" any code in order
to use the TF backend. Any code that works with Theano-Keras will run on
TF-Keras.

**Comment by shivarajugowda** [comment]
I am working on this. I feel it is worthwhile endeavor since Keras will benefit from the MXNET’s multi-device/multi-node support and MXNET will get broader exposure along with some updates to its API capabilities. Keras's well written backend interface and tests are very helpful so far, I have been able to quickly get off the ground and handle the low hanging APIs. I am simultaneously working with MXNET to add in the missing ones. I would say around 25% of the APIs are covered till now. You can monitor the progress at these following locations. 

Keras : https://github.com/shivarajugowda/keras
MXNET: https://github.com/dmlc/mxnet/issues/4173

**Comment by pGit1** [comment]
Keras with MXNET backend...

Multi-GPU - Check
Insanely Fast - Check

**Comment by smolix** [comment]
@pGit1 
Multi-Machine - Check
Linear Scaling to 256 GPUs - Check

**Comment by fchollet** [comment]
@shivarajugowda that's very cool. Feel free to close this issue and open a new one to track progress on this new backend and to raise any issues you encounter in the process.

**Comment by shivarajugowda** [comment]
@fchollet thanks, will do that.

**Comment by pGit1** [comment]
@smolix 

Thats just insanity... 👍

**Comment by yajiedesign** [comment]
ok.the mxnet backend now can run,.download in https://github.com/dmlc/keras

**Comment by pGit1** [comment]
@yajiedesign 

Can you please provide 2 examples?

1. Simple ConvNet on MNIST utilizing multiple GPUs and (2 or more). Hopefully this can demonstrate the linear scaling across GPUs on a single machine. 

2. Using a pre-trained model with Mxnet backend.
